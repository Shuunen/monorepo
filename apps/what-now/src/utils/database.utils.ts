import { dateIso10, nbPercentMax, nbSpacesIndent, Result, slugify, toastError, toastSuccess } from '@monorepo/utils'
import { Client, Databases, type Models, Query } from 'appwrite'
import type { AppWriteTask, Task } from '../types'
import { logger } from './logger.utils'
import { state } from './state.utils'

const client = new Client()
const database = new Databases(client)
client.setEndpoint('https://cloud.appwrite.io/v1').setProject('what-now')

export type AppWriteTaskModel = AppWriteTask & Models.Document

export const uuidMaxLength = 36

/**
 * Convert a task from the database to a task in the app
 * @param task the db task to convert
 * @returns the app task
 */
export function modelToLocalTask(task: AppWriteTaskModel) {
  return {
    completedOn: task['completed-on'],
    id: task.$id,
    isDone: task.done,
    minutes: task.minutes,
    name: task.name,
    once: task.once,
    reason: task.reason ?? undefined,
  } satisfies Task
}

/**
 * Create task for database insertion
 * @param model - The uploaded task
 * @returns AppWrite task format
 */
export function modelToRemoteTask(model: AppWriteTaskModel) {
  return {
    'completed-on': model['completed-on'],
    done: model.done,
    minutes: model.minutes,
    name: model.name,
    once: model.once,
    reason: model.reason,
  } satisfies AppWriteTask
}

/**
 * Convert a task from the app to the task format used in the database
 * @param task the app task to convert
 * @returns the db task
 */
export function localToRemoteTask(task: Readonly<Task>) {
  return {
    'completed-on': task.completedOn,
    done: task.isDone,
    minutes: task.minutes,
    name: task.name,
    once: task.once,
    reason: task.reason ?? '',
  } satisfies AppWriteTask
}

/**
 * Add a task to the database
 * @param data the task to add
 * @returns the result of the operation
 */
export function addTask(data: Readonly<AppWriteTask>) {
  const id = slugify(data.name).slice(0, uuidMaxLength)
  return Result.trySafe(database.createDocument(state.apiDatabase, state.apiCollection, id, data))
}

/**
 * Update a task in the database
 * @param task the task to update
 * @returns the result of the operation
 */
export function updateTask(task: Readonly<Task>) {
  const data = localToRemoteTask(task)
  return Result.trySafe(database.updateDocument<AppWriteTaskModel>(state.apiDatabase, state.apiCollection, task.id, data))
}

/**
 * Get all tasks from the database
 * @returns the result of the operation
 */
export async function getTasks() {
  const result = await Result.trySafe(database.listDocuments<AppWriteTaskModel>(state.apiDatabase, state.apiCollection, [Query.limit(nbPercentMax)]))
  if (!result.ok) return result
  const tasks = result.value.documents.map<Task>(task => modelToLocalTask(task))
  logger.info(`found ${tasks.length} tasks on db`, tasks)
  return Result.ok(tasks)
}

/* c8 ignore start */
/**
 * Download the data from the database
 * @returns the result of the operation
 */
export async function downloadData() {
  const result = await Result.trySafe(database.listDocuments<AppWriteTaskModel>(state.apiDatabase, state.apiCollection, [Query.limit(nbPercentMax)]))
  if (!result.ok) {
    toastError('Failed to download data')
    logger.error('failed to download data', result.error)
    return
  }
  const json = JSON.stringify(result.value.documents, undefined, nbSpacesIndent)
  const blob = new Blob([json], { type: 'application/json' })
  const url = URL.createObjectURL(blob)
  const link = document.createElement('a')
  link.href = url
  link.download = `${dateIso10()}_what-now_tasks.json`
  document.body.append(link)
  link.click()
  link.remove()
  URL.revokeObjectURL(url)
  toastSuccess('Data downloaded')
}
