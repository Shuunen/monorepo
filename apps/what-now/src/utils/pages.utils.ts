import type { FloatingMenuAction } from '@shuunen/components'
import { CircleQuestionMarkIcon, HomeIcon, SettingsIcon } from 'lucide-react'
import { useNavigate } from 'react-router-dom'

/**
 * Returns a filtered list of navigation actions, excluding any whose name includes the given string.
 * Each action contains a `handleClick` function for navigation, an `icon` component, and a `name`.
 * @returns An array of action objects excluding those whose names include the `except` string.
 */
export function useActions() {
  const navigate = useNavigate()
  const currentPath = globalThis.location.pathname
  // oxlint-disable-next-line require-param
  function getAction(icon: FloatingMenuAction['icon'], name: FloatingMenuAction['name'], path: string) {
    return {
      disabled: path === currentPath,
      handleClick: () => navigate(path),
      icon,
      name,
    } satisfies FloatingMenuAction
  }
  return [getAction(HomeIcon, 'Tasks', '/'), getAction(SettingsIcon, 'Settings', '/settings'), getAction(CircleQuestionMarkIcon, 'About', '/about')]
}
