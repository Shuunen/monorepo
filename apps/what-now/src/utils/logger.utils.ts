/* c8 ignore start */
import { isBrowserEnvironment, Logger } from '@shuunen/shuutils'
import { state } from './state.utils'

function stuffToMessage(...stuff: unknown[]) {
  return stuff
    .map(thing => {
      if (typeof thing === 'string') return thing
      if (typeof thing === 'object') return JSON.stringify(thing)
      return String(thing)
    })
    .join(', ')
}
class CustomLogger extends Logger {
  public override error(...stuff: unknown[]) {
    super.error(...stuff)
    state.showErrorToast = stuffToMessage(...stuff)
  }
}

const logger = new CustomLogger({ minimumLevel: '3-info', willOutputToConsole: isBrowserEnvironment() })

export { logger, stuffToMessage }
