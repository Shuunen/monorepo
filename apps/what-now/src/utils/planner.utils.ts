import { Result } from '@shuunen/utils'
import type { Task } from '../types'
import { updateTask } from './database.utils'
import { daysRecurrence, daysSinceCompletion } from './tasks.utils'

export const weekDays = ['Mon W1', 'Tue W1', 'Wed W1', 'Thu W1', 'Fri W1', 'Sat W1', 'Sun W1', 'Mon W2', 'Tue W2', 'Wed W2', 'Thu W2', 'Fri W2', 'Sat W2', 'Sun W2']
export const daysInWeek = 14
// oxlint-disable-next-line no-magic-numbers
export const allDayIndices = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]
export const dailyRecurrence = 1
export const twoDayRecurrence = 2
export const threeDayRecurrence = 3
export const weeklyRecurrence = 7

// Constants for recurrence patterns
export const dayFrequency = 1
export const weekFrequency = 7
export const biWeeklyFrequency = 14
export const triWeeklyFrequency = 21
export const monthlyFrequency = 28

// Valid frequency options in ascending order (higher frequency = lower number)
// oxlint-disable-next-line no-magic-numbers
export const validFrequencies = [dayFrequency, 2, 3, 4, 5, 6, weekFrequency, biWeeklyFrequency, triWeeklyFrequency, monthlyFrequency]

/**
 * Converts days recurrence to frequency string
 * @param days - Number of days between recurrences
 * @returns Frequency string for the task's once property
 */
export function daysToFrequencyString(days: number): string {
  if (days === dayFrequency) return 'day'
  if (days === weekFrequency) return 'week'
  if (days === biWeeklyFrequency) return '2-weeks'
  if (days === triWeeklyFrequency) return '3-weeks'
  if (days === monthlyFrequency) return '4-weeks'
  return `${days}-days`
}

/**
 * Gets the next higher frequency (lower days)
 * @param currentDays - Current recurrence in days
 * @returns Next higher frequency or undefined if already at maximum
 */
export function getHigherFrequency(currentDays: number): number | undefined {
  const currentIndex = validFrequencies.indexOf(currentDays)
  if (currentIndex <= 0) return undefined // Already at highest frequency (daily)
  return validFrequencies[currentIndex - 1]
}

/**
 * Gets the next lower frequency (higher days)
 * @param currentDays - Current recurrence in days
 * @returns Next lower frequency or undefined if already at minimum
 */
export function getLowerFrequency(currentDays: number): number | undefined {
  const currentIndex = validFrequencies.indexOf(currentDays)
  const notFound = -1
  if (currentIndex === notFound || currentIndex >= validFrequencies.length - 1) return undefined
  return validFrequencies[currentIndex + 1]
}

/**
 * Gets the effective recurrence for a task (considering modifications)
 * @param task - The task
 * @param modifications - Current modifications
 * @returns The effective recurrence in days
 */
export function getEffectiveRecurrence(task: Task, modifications: Record<string, number>): number {
  return modifications[task.id] ?? daysRecurrence(task)
}

/**
 * Gets a color class based on task recurrence for visual distinction
 * @param task - The task to get color for
 * @param modifications - Current modifications (optional)
 * @returns CSS classes for styling the task
 */
export function getTaskColor(task: Task, modifications: Record<string, number> = {}): string {
  const recurrence = getEffectiveRecurrence(task, modifications)
  if (recurrence === dailyRecurrence) return 'bg-red-900/10 border-red-500/30 text-red-300'
  if (recurrence === twoDayRecurrence) return 'bg-orange-900/10 border-orange-500/30 text-orange-300'
  if (recurrence === threeDayRecurrence) return 'bg-blue-900/10 border-blue-500/30 text-blue-300'
  if (recurrence >= weeklyRecurrence) return 'bg-green-900/10 border-green-500/30 text-green-300'
  return 'bg-gray-800/30 border-gray-600/30 text-gray-300'
}

/**
 * Calculates which days of the current two weeks a task should appear based on recurrence
 * @param daysSinceComplete - Days since task was last completed
 * @param recurrence - Task recurrence in days
 * @returns Array of day indices where task should appear
 */
function calculateWeeklyTaskDays(daysSinceComplete: number, recurrence: number): number[] {
  const taskDays: number[] = []
  const today = new Date()
  const sundayToMondayOffset = 6
  const daysInSingleWeekConstant = 7
  const currentDayOfWeek = (today.getDay() + sundayToMondayOffset) % daysInSingleWeekConstant
  for (let dayIndex = 0; dayIndex < daysInWeek; dayIndex += 1) {
    const daysFromToday = dayIndex - currentDayOfWeek
    const totalDaysFromLastCompletion = daysSinceComplete + daysFromToday
    if (totalDaysFromLastCompletion >= recurrence && totalDaysFromLastCompletion % recurrence === 0) taskDays.push(dayIndex)
  }
  return taskDays
}

/**
 * Determines which days of the two weeks a task should appear based on its effective recurrence pattern
 * @param task - The task to analyze
 * @param modifications - Current modifications to tasks
 * @returns Array of day indices (0-13) where the task should appear
 */
export function getTaskDaysWithModifications(task: Task, modifications: Record<string, number>): number[] {
  const recurrence = getEffectiveRecurrence(task, modifications)
  if (recurrence === 0 || task.once === 'yes') return []
  if (recurrence === dailyRecurrence) return allDayIndices
  const daysSinceComplete = daysSinceCompletion(task)
  return calculateWeeklyTaskDays(daysSinceComplete, recurrence)
}

/**
 * Creates the task distribution data structure
 * @param tasks - Array of tasks to distribute
 * @param modifications - Current modifications to tasks
 * @returns Record mapping day indices to task arrays
 */
export function createTaskDistribution(tasks: Task[], modifications: Record<string, number> = {}) {
  // Create a map of day index to tasks for that day
  const tasksByDay: Record<number, Task[]> = {}
  for (let dayIndex = 0; dayIndex < daysInWeek; dayIndex += 1) tasksByDay[dayIndex] = []
  // Distribute tasks across days based on their effective recurrence
  for (const task of tasks) {
    const taskDays = getTaskDaysWithModifications(task, modifications)
    for (const dayIndex of taskDays) tasksByDay[dayIndex].push(task)
  }
  return tasksByDay
}

/**
 * Creates an update promise for a single task modification
 * @param taskId - ID of the task to update
 * @param newDays - New recurrence in days
 * @param tasks - Array of all tasks
 * @returns Promise for updating the task
 */
function createTaskUpdatePromise(taskId: string, newDays: number, tasks: Task[]) {
  const task = tasks.find(currentTask => currentTask.id === taskId)
  if (!task) return Promise.resolve(Result.error(`failed to find task ${taskId}`))
  const frequencyString = daysToFrequencyString(newDays)
  return updateTask({ ...task, once: frequencyString })
}

/**
 * Creates a date update promise for a task
 * @param taskId - Task ID to update
 * @param newCompletedOn - New completion date
 * @param tasks - Array of all tasks
 * @returns Promise that resolves to update result
 */
function createTaskDateUpdatePromise(taskId: string, newCompletedOn: string, tasks: Task[]) {
  const task = tasks.find(currentTask => currentTask.id === taskId)
  if (!task) return Promise.resolve(Result.error(`task with id ${taskId} not found`))
  return updateTask({ ...task, completedOn: newCompletedOn })
}

/**
 * Processes all update promises and checks if they were successful
 * @param updatePromises - Array of update promises
 * @returns Promise that resolves to Result indicating success or failure
 */
async function processUpdatePromises(updatePromises: Promise<unknown>[]) {
  const results = await Promise.allSettled(updatePromises)
  const allSuccessful = results.every(result => {
    if (result.status !== 'fulfilled') return false
    const value = result.value as { ok: boolean }
    return value?.ok
  })
  if (!allSuccessful) return Result.error('failed to save task modifications')
  return Result.ok('successfully saved task modifications')
}

/**
 * Handles saving task frequency modifications
 * @param frequencyModifications - Record of task ID to new frequency in days
 * @param dateModifications - Record of task ID to new completion date
 * @param tasks - Array of all tasks
 * @returns Promise that resolves when save is complete
 */
export function saveTaskModifications(frequencyModifications: Record<string, number>, dateModifications: Record<string, string>, tasks: Task[]) {
  const updatePromises: Promise<unknown>[] = []

  // Handle frequency modifications
  const frequencyEntries = Object.entries(frequencyModifications)
  const frequencyPromises = frequencyEntries.map(([taskId, newDays]) => createTaskUpdatePromise(taskId, newDays, tasks))
  updatePromises.push(...frequencyPromises)

  // Handle date modifications
  const dateEntries = Object.entries(dateModifications)
  const datePromises = dateEntries.map(([taskId, newCompletedOn]) => createTaskDateUpdatePromise(taskId, newCompletedOn, tasks))
  updatePromises.push(...datePromises)

  return processUpdatePromises(updatePromises)
}
