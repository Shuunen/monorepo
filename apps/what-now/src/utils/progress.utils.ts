import { state } from './state.utils'
import { isTaskActive, minutesRemaining } from './tasks.utils'

export function webhookPayload(progress = 0) {
  const activeTasks = state.tasks.filter(task => isTaskActive(task))
  const firstTask = activeTasks[0]
  const minutes = minutesRemaining(activeTasks)
  return `progress=${progress}&remaining=${minutes}&nextTask=${firstTask?.name}`
}
