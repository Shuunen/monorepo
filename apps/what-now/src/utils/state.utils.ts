import { createState, storage } from '@monorepo/utils'
import type { Task } from '../types'

const tasks: Task[] = []

export const { state, watchState } = createState(
  {
    apiCollection: '',
    apiDatabase: '',
    isLoading: false,
    /** true if the app is ready to use */
    isSetup: false,
    showErrorToast: '',
    showInfoToast: '',
    statusError: '',
    statusInfo: 'Loading, please wait...',
    statusProgress: '',
    /** the last list of tasks fetched */
    tasks,
    /** timestamp of the last time tasks were fetched, in milliseconds */
    tasksTimestamp: 0,
    /** the webhook URL to connect the app with external services */
    webhook: '',
  },
  storage,
  /* c8 ignore next */ ['apiDatabase', 'apiCollection', 'tasks', 'tasksTimestamp', 'webhook'],
)

export type CredentialField = keyof Pick<typeof state, 'apiCollection' | 'apiDatabase' | 'webhook'>
