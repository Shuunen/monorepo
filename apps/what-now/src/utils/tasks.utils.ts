import { dateIso10, daysAgoIso10, nbBefore, nbDaysInMonth, nbDaysInWeek, nbDaysInYear, nbMsInDay, nbMsInMinute, Result, readableTimeAgo } from '@shuunen/utils'
import type { Task } from '../types'
import { getTasks, updateTask } from './database.utils'
import { logger } from './logger.utils'
import { state } from './state.utils'

const recurrenceRegex = /(?<quantity>\d{1,3})?-?(?<unit>day|month|week|year)/u

/**
 * Calculates the total number of minutes remaining for all active tasks in the provided array.
 * @param tasks - An array of `Task` objects to evaluate.
 * @returns The total minutes remaining for all active tasks.
 */
export function minutesRemaining(tasks: Task[]) {
  let minutes = 0
  for (const task of tasks) if (isTaskActive(task)) minutes += task.minutes
  return minutes
}

export function daysRecurrence(task: Task) {
  const matches = recurrenceRegex.exec(task.once)
  if (matches === null) return 0
  const quantity = matches.groups?.quantity ?? '1'
  const unit = matches.groups?.unit as 'day' | 'month' | 'week' | 'year'
  const number = Number.parseInt(quantity, 10)
  if (unit === 'day') return number
  if (unit === 'week') return number * nbDaysInWeek
  if (unit === 'month') return number * nbDaysInMonth
  return number * nbDaysInYear
}

export function daysSinceCompletion(task: Task) {
  const today = dateIso10(new Date())
  const todayTimestamp = new Date(today).getTime()
  const completedOnTimestamp = new Date(task.completedOn).getTime()
  return (todayTimestamp - completedOnTimestamp) / nbMsInDay
}

export function completeTask(task: Task) {
  task.completedOn = dateIso10(new Date()) // task is complete for today
  task.isDone = task.once === 'yes' // but it also can be done totally if it was a one time job
  return updateTask(task)
}

export function unCompleteTask(task: Task) {
  // to un-complete, need to put the last completed on just before the required number of days
  task.completedOn = daysAgoIso10(daysRecurrence(task))
  task.isDone = false
  return updateTask(task)
}

export function isTaskActive(task: Task, shouldIncludeCompletedToday = false) {
  if (task.isDone) return false
  if (task.completedOn === '' || task.once === 'yes') return true
  const recurrence = daysRecurrence(task)
  const days = daysSinceCompletion(task)
  const isActive = (shouldIncludeCompletedToday && days === 0) || days >= recurrence
  return isActive
}

export function byActive(taskA: Task, taskB: Task) {
  const isA = isTaskActive(taskA)
  const isB = isTaskActive(taskB)
  if (isA && !isB) return -1
  if (!isA && isB) return 1
  return 0
}

export async function fetchList() {
  logger.info('fetch list')
  state.statusInfo = 'Loading tasks, please wait...'
  const result = await getTasks()
  if (!result.ok) return result
  state.statusInfo = ''
  state.tasksTimestamp = Date.now()
  const list = result.value.filter(task => isTaskActive(task, true)).sort(byActive)
  return Result.ok(list)
}

export function isDataOlderThan(milliseconds: number) {
  if (!state.tasksTimestamp) return true
  const age = Date.now() - state.tasksTimestamp
  const minutes = Math.round(age / nbMsInMinute)
  /* c8 ignore next */
  if (minutes > 0) logger.info('last activity', minutes, 'minute(s) ago')
  return age >= milliseconds
}

export async function loadTasks() {
  if (!state.isSetup) return Result.error('not setup, cannot load tasks')
  if (state.tasks.length > 0 && !isDataOlderThan(Number(nbMsInMinute))) return Result.ok(`tasks are fresh (${readableTimeAgo(Date.now() - state.tasksTimestamp)})`)
  const result = await fetchList()
  if (!result.ok) return result
  logger.info('found', result.value.length, 'task(s)')
  state.isLoading = false
  state.tasks = result.value
  return Result.ok('tasks loaded')
}

export function dispatchTask(task: Task, index = 0) {
  if (['day', 'yes'].includes(task.once)) return Result.error(task.once === 'yes' ? 'one-time task, cannot dispatch' : 'daily task, nothing to dispatch')
  const delay = daysRecurrence(task)
  const position = index % delay
  const completionDate = daysAgoIso10(nbBefore * position + delay)
  if (completionDate === task.completedOn) return Result.error('task already dispatched')
  task.completedOn = completionDate
  return Result.ok(task) // task dispatched
}

export function dispatchTasks(tasks: Task[]) {
  logger.info('dispatch tasks...')
  return tasks.map(task => dispatchTask(task))
}

export async function dispatchTasksAndUpdate(tasks: Task[]) {
  logger.info('dispatch tasks and update...')
  const taskUpdates = dispatchTasks(tasks).map(result => {
    if (result.ok) return updateTask(result.value)
    return Promise.resolve()
  })
  await Promise.all(taskUpdates)
}

export function toggleComplete(task: Task) {
  if (isTaskActive(task)) return completeTask(task)
  return unCompleteTask(task)
}

/**
 * Create a task mock with default values
 * @param fields - fields to override
 * @returns a task mock
 */
export function taskMock(fields: Partial<Task> = {}): Task {
  const { completedOn = daysAgoIso10(0), id = 'id-123', isDone = false, minutes = 20, name = 'a super task', once = 'day', reason } = { ...fields }
  return { completedOn, id, isDone, minutes, name, once, reason } satisfies Task
}
