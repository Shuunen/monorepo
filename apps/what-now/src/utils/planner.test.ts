import { Result } from '@shuunen/shuutils'
import type { Models } from 'appwrite'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import type { AppWriteTask, Task } from '../types'
import * as databaseUtils from './database.utils'
import { allDayIndices, createTaskDistribution, daysInWeek, daysToFrequencyString, getEffectiveRecurrence, getHigherFrequency, getLowerFrequency, getTaskColor, getTaskDaysWithModifications, saveTaskModifications, validFrequencies, weekDays } from './planner.utils'
import * as tasksUtils from './tasks.utils'

vi.mock('./database.utils')
vi.mock('./tasks.utils')

type AppWriteTaskModel = AppWriteTask & Models.Document

const mockTask: Task = {
  completedOn: '2023-01-01T10:00:00Z',
  id: 'task-1',
  isDone: false,
  minutes: 30,
  name: 'Test Task',
  once: 'week',
}

const mockAppWriteTask: AppWriteTaskModel = {
  $collectionId: 'collection-id',
  $createdAt: '2023-01-01T10:00:00Z',
  $databaseId: 'database-id',
  $id: 'task-1',
  $permissions: [],
  $sequence: 1,
  $updatedAt: '2023-01-01T10:00:00Z',
  'completed-on': '2023-01-01T10:00:00Z',
  done: false,
  minutes: 30,
  name: 'Test Task',
  once: '2-weeks',
}

describe('planner utils', () => {
  beforeEach(() => {
    vi.clearAllMocks()
    // Mock current date for consistent testing
    vi.setSystemTime(new Date('2023-01-08T10:00:00Z')) // Sunday
  })

  it('weekDays A should contain all fourteen days', () => {
    expect(weekDays).toMatchSnapshot()
  })

  it('daysInWeek A should equal 14', () => {
    expect(daysInWeek).toBe(14)
  })

  it('allDayIndices A should contain indices for all days', () => {
    expect(allDayIndices).toMatchSnapshot()
  })

  it('validFrequencies A should be in ascending order', () => {
    expect(validFrequencies).toMatchSnapshot()
  })

  it('daysToFrequencyString A should convert 1 to day', () => {
    expect(daysToFrequencyString(1)).toBe('day')
  })

  it('daysToFrequencyString B should convert 7 to week', () => {
    expect(daysToFrequencyString(7)).toBe('week')
  })

  it('daysToFrequencyString C should convert 14 to 2-weeks', () => {
    expect(daysToFrequencyString(14)).toBe('2-weeks')
  })

  it('daysToFrequencyString D should convert 21 to 3-weeks', () => {
    expect(daysToFrequencyString(21)).toBe('3-weeks')
  })

  it('daysToFrequencyString E should convert 28 to 4-weeks', () => {
    expect(daysToFrequencyString(28)).toBe('4-weeks')
  })

  it('daysToFrequencyString F should convert other numbers to X-days format', () => {
    expect(daysToFrequencyString(5)).toBe('5-days')
  })

  it('getHigherFrequency A should return lower number for higher frequency', () => {
    expect(getHigherFrequency(7)).toBe(6)
  })

  it('getHigherFrequency B should return undefined for daily frequency', () => {
    expect(getHigherFrequency(1)).toBeUndefined()
  })

  it('getHigherFrequency C should return undefined for invalid frequency', () => {
    expect(getHigherFrequency(10)).toBeUndefined()
  })

  it('getLowerFrequency A should return higher number for lower frequency', () => {
    expect(getLowerFrequency(7)).toBe(14)
  })

  it('getLowerFrequency B should return undefined for monthly frequency', () => {
    expect(getLowerFrequency(28)).toBeUndefined()
  })

  it('getLowerFrequency C should return undefined for invalid frequency', () => {
    expect(getLowerFrequency(10)).toBeUndefined()
  })

  it('getEffectiveRecurrence A should return modification when present', () => {
    const modifications = { 'task-1': 14 }
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(7)
    expect(getEffectiveRecurrence(mockTask, modifications)).toBe(14)
  })

  it('getEffectiveRecurrence B should return original recurrence when no modification', () => {
    const modifications = {}
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(7)
    expect(getEffectiveRecurrence(mockTask, modifications)).toBe(7)
  })

  it('getTaskColor A should return red for daily tasks', () => {
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(1)
    expect(getTaskColor(mockTask)).toContain('bg-red-900/10')
  })

  it('getTaskColor B should return orange for 2-day tasks', () => {
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(2)
    expect(getTaskColor(mockTask)).toContain('bg-orange-900/10')
  })

  it('getTaskColor C should return blue for 3-day tasks', () => {
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(3)
    expect(getTaskColor(mockTask)).toContain('bg-blue-900/10')
  })

  it('getTaskColor D should return green for weekly or longer tasks', () => {
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(7)
    expect(getTaskColor(mockTask)).toContain('bg-green-900/10')
  })

  it('getTaskColor E should return gray for other tasks', () => {
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(0)
    expect(getTaskColor(mockTask)).toContain('bg-gray-800/30')
  })

  it('getTaskColor F should use modifications when provided', () => {
    const modifications = { 'task-1': 1 }
    expect(getTaskColor(mockTask, modifications)).toContain('bg-red-900/10')
  })

  it('getTaskDaysWithModifications A should return empty array for one-time tasks', () => {
    const oneTimeTask = { ...mockTask, once: 'yes' }
    expect(getTaskDaysWithModifications(oneTimeTask, {})).toEqual([])
  })

  it('getTaskDaysWithModifications B should return empty array for zero recurrence', () => {
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(0)
    expect(getTaskDaysWithModifications(mockTask, {})).toEqual([])
  })

  it('getTaskDaysWithModifications C should return all days for daily tasks', () => {
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(1)
    expect(getTaskDaysWithModifications(mockTask, {})).toEqual(allDayIndices)
  })

  it('getTaskDaysWithModifications D should calculate weekly task days correctly', () => {
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(7)
    vi.mocked(tasksUtils.daysSinceCompletion).mockReturnValue(7)
    const result = getTaskDaysWithModifications(mockTask, {})
    expect(result).toMatchInlineSnapshot(`
      [
        6,
        13,
      ]
    `)
  })

  it('createTaskDistribution A should create empty distribution for empty task list', () => {
    const result = createTaskDistribution([])
    expect(result).toMatchSnapshot()
  })

  it('createTaskDistribution B should distribute daily task to all days', () => {
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(1)
    const result = createTaskDistribution([mockTask])
    expect(Object.values(result).every(dayTasks => dayTasks.length === 1)).toBe(true)
  })

  it('createTaskDistribution C should distribute weekly task to specific days', () => {
    vi.mocked(tasksUtils.daysRecurrence).mockReturnValue(7)
    vi.mocked(tasksUtils.daysSinceCompletion).mockReturnValue(7)
    const result = createTaskDistribution([mockTask])
    expect(result[6]).toHaveLength(1)
    expect(result[0]).toHaveLength(0)
  })

  it('createTaskDistribution D should use modifications when provided', () => {
    const modifications = { 'task-1': 1 }
    const result = createTaskDistribution([mockTask], modifications)
    expect(Object.values(result).every(dayTasks => dayTasks.length === 1)).toBe(true)
  })

  it('saveTaskModifications A should save all modifications successfully', async () => {
    const modifications = { 'task-1': 14 }
    const dateModifications = {}
    const tasks = [mockTask]
    vi.mocked(databaseUtils.updateTask).mockResolvedValue(Result.ok(mockAppWriteTask))
    const result = await saveTaskModifications(modifications, dateModifications, tasks)
    expect(result.ok).toBe(true)
    expect(databaseUtils.updateTask).toHaveBeenCalledWith({
      ...mockTask,
      once: '2-weeks',
    })
  })

  it('saveTaskModifications B should return error when task not found', async () => {
    const modifications = { 'nonexistent-task': 14 }
    const dateModifications = {}
    const tasks = [mockTask]
    const result = await saveTaskModifications(modifications, dateModifications, tasks)
    expect(result.ok).toBe(false)
  })

  it('saveTaskModifications C should return error when update fails', async () => {
    const modifications = { 'task-1': 14 }
    const dateModifications = {}
    const tasks = [mockTask]
    vi.mocked(databaseUtils.updateTask).mockResolvedValue(Result.error('update failed'))
    const result = await saveTaskModifications(modifications, dateModifications, tasks)
    expect(result.ok).toBe(false)
  })

  it('saveTaskModifications D should handle mixed success and failure', async () => {
    const modifications = { 'task-1': 14, 'task-2': 7 }
    const dateModifications = {}
    const task2 = { ...mockTask, id: 'task-2' }
    const tasks = [mockTask, task2]
    vi.mocked(databaseUtils.updateTask).mockResolvedValueOnce(Result.ok(mockAppWriteTask)).mockResolvedValueOnce(Result.error('failure'))
    const result = await saveTaskModifications(modifications, dateModifications, tasks)
    expect(result.ok).toBe(false)
  })

  it('saveTaskModifications E should save date modifications successfully', async () => {
    const modifications = {}
    const dateModifications = { 'task-1': '2023-10-01T00:00:00.000Z' }
    const tasks = [mockTask]
    vi.mocked(databaseUtils.updateTask).mockResolvedValue(Result.ok(mockAppWriteTask))
    const result = await saveTaskModifications(modifications, dateModifications, tasks)
    expect(result.ok).toBe(true)
    expect(databaseUtils.updateTask).toHaveBeenCalledWith({
      ...mockTask,
      completedOn: '2023-10-01T00:00:00.000Z',
    })
  })

  it('saveTaskModifications F should save both frequency and date modifications', async () => {
    const modifications = { 'task-1': 7 }
    const dateModifications = { 'task-1': '2023-10-01T00:00:00.000Z' }
    const tasks = [mockTask]
    vi.mocked(databaseUtils.updateTask).mockResolvedValue(Result.ok(mockAppWriteTask))
    const result = await saveTaskModifications(modifications, dateModifications, tasks)
    expect(result.ok).toBe(true)
    expect(databaseUtils.updateTask).toHaveBeenCalledTimes(2) // Should be called twice: once for frequency, once for date
  })

  it('saveTaskModifications G should return error when date modification task not found', async () => {
    const modifications = {}
    const dateModifications = { 'nonexistent-task': '2023-10-01T00:00:00.000Z' }
    const tasks = [mockTask]
    const result = await saveTaskModifications(modifications, dateModifications, tasks)
    expect(result.ok).toBe(false)
  })

  it('saveTaskModifications H should handle promise rejection', async () => {
    const modifications = { 'task-1': 7 }
    const dateModifications = {}
    const tasks = [mockTask]
    vi.mocked(databaseUtils.updateTask).mockRejectedValue(new Error('network error'))
    const result = await saveTaskModifications(modifications, dateModifications, tasks)
    expect(result.ok).toBe(false)
  })
})
