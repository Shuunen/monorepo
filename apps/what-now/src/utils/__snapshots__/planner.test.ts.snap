// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`planner utils > allDayIndices A should contain indices for all days 1`] = `
[
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
]
`;

exports[`planner utils > createTaskDistribution A should create empty distribution for empty task list 1`] = `
{
  "0": [],
  "1": [],
  "10": [],
  "11": [],
  "12": [],
  "13": [],
  "2": [],
  "3": [],
  "4": [],
  "5": [],
  "6": [],
  "7": [],
  "8": [],
  "9": [],
}
`;

exports[`planner utils > validFrequencies A should be in ascending order 1`] = `
[
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  14,
  21,
  28,
]
`;

exports[`planner utils > weekDays A should contain all fourteen days 1`] = `
[
  "Mon W1",
  "Tue W1",
  "Wed W1",
  "Thu W1",
  "Fri W1",
  "Sat W1",
  "Sun W1",
  "Mon W2",
  "Tue W2",
  "Wed W2",
  "Thu W2",
  "Fri W2",
  "Sat W2",
  "Sun W2",
]
`;
